### PR TITLE
Sphinx doc

### PR DESCRIPTION
**Pull request by pmaybank**
Set up infrastructure for generating HTML doc and start on a CK Primitives Guide.

### CODE REVIEW

**Review by geyyer — CHANGES_REQUESTED** [review]
Hi @pmaybank, thanks for adding this! Overall the PR looks like exactly what we need! Could you address the comments? Do you think we should add these docs to the new "docs" folder or add to the existing "doc" one? Which dependencies are we going to need for docs generation? Do you think we should add them to our docker images?

File: .gitignore [hunk 1]
```diff
@@ -47,3 +47,7 @@ build*
 # GDB temporary files
 .gdb_history
 install.dir*
+
+# directories containing generated documentation
+docs/source/_build/
+docs/docBin/
```
[the remark this answers is not in the document]
Comment (pmaybank) [reply .gitignore]: Done

File: docs/source/Disclaimer.rst [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
...D ASSUMES NO RESPONSIBILITY FOR ANY INACCURACIES, ERRORS, OR OMISSIONS THAT MAY APPEAR IN THIS INFORMATION. AMD SPECIFICALLY DISCLAIMS ANY IMPLIED WARRANTIES OF NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR ANY PARTICULAR PURPOSE. IN NO EVENT WILL AMD BE LIABLE TO ANY PERSON FOR ANY RELIANCE, DIRECT, INDIRECT, SPECIAL, OR OTHER CONSEQUENTIAL DAMAGES ARISING FROM THE USE OF ANY INFORMATION CONTAINED HEREIN, EVEN IF AMD IS EXPRESSLY ADVISED OF THE POSSIBILITY OF SUCH DAMAGES. AMD, the AMD Arrow logo, Radeon, Ryzen, Epyc, and combinations thereof are trademarks of Advanced Micro Devices, Inc. Other product names used in this publication are for identification purposes only and may be trademarks of their respective companies. Google(R)  is a registered trademark of Google LLC. PCIe(R) is a registered trademark of PCI-SIG Corporation. Linux(R) is the registered trademark of Linus Torvalds in the U.S. and other countries. Ubuntu(R) and the Ubuntu logo are registered trademarks of Canonical Ltd. Other product names used in this publication are for identification purposes only and may be trademarks of their respective companies. (C)2022 Advanced Micro Devices, Inc. All rights reserved.
```
[the remark this answers is not in the document]
Comment (pmaybank) [reply docs/source/Disclaimer.rst]: Done

File: docs/source/conf.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
+"""Copyright (C) 2016-2023 Advanced Micro Devices, Inc. All rights reserved.
```
[the remark this answers is not in the document]
Comment (pmaybank) [reply docs/source/conf.py]: Done

File: docs/source/conf.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+#
+# source_suffix = ['.rst', '.md']
+source_suffix = '.rst'
+
+# The master toctree document.
+master_doc = 'index'
+
+# General information about the project.
+project = u'Composable Kernel (CK)'
+copyright = u'2016-2023, Advanced Micro Devices'
```
[the remark this answers is not in the document]
Comment (pmaybank) [reply docs/source/conf.py]: Done

**Comment by pmaybank** [comment]
> Hi @pmaybank, thanks for adding this! Overall the PR looks like exactly what we need! ...

Thanks @geyyer for taking the time to look at this PR.  As a general comment before answering your questions, I based this work on what has been done for rocBLAS and rocFFT repositories.  I think following what has been done in those repositories is the simplest way forward, unless that are strong reasons for doing something different.  For example, this means using reStructuredText by default rather than Markdown.

Also, I was not aware of the doc directory when I created this PR as I had been working on a branch created before you created the doc directory.  Should have looked a bit more carefully as I may have done things a bit differently if I had known that doc was already there.

Re. your general questions,

> Do you think we should add these docs to the new "docs" folder or add to the existing "doc" one? 

rocBLAS and rocFFT both have "docs" directories.  I am inclined to follow what is done there.

> Which dependencies are we going to need for docs generation? 

Doxygen, Sphinx, and optionally CMake, I think.  Parts of the following blog may be helpful for getting dependencies set-up - https://devblogs.microsoft.com/cppblog/clear-functional-c-documentation-with-sphinx-breathe-doxygen-cmake/ .

> Do you think we should add them to our docker images?

I don't have much experience of generating / using docker images, so don't know.  Would be good to discuss.  I don't know if this is related, but I would like there to be some kind of pipeline for generating HTML that gets triggered whenever anyone pushes to remote so that latest doc can be easily accessed in a web browser.

Would be interested to discuss further in case I have missed anything.

File: docs/source/conf.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+breathe_default_project = "CK"
+
+# Add any paths that contain templates here, relative to this directory.
+templates_path = ['_templates']
+
+# The suffix(es) of source filenames.
+# You can specify multiple suffix as a list of string:
+#
+# source_suffix = ['.rst', '.md']
+source_suffix = '.rst'
```
[the remark this answers is not in the document]
Comment (geyyer) [reply docs/source/conf.py:73]: I'm totally fine with just one "language / markup" for writing developer guides, let's stick to .rst.

**Comment by geyyer** [comment]
@pmaybank, thanks for the editions and adding the existing docs!

> Do you think we should add these docs to the new "docs" folder or add to the existing "doc" one?
> 
> rocBLAS and rocFFT both have "docs" directories. I am inclined to follow what is done there.

Hmm, indeed rocBLAS and rocFFT have "docs", but MIOpen library has a "doc" dir https://github.com/ROCmSoftwarePlatform/MIOpen/tree/develop/doc. @zjing14, @asroy, what do you think is our plan to get a consistent naming within ROCm?

> Do you think we should add them to our docker images?
> 
> I don't have much experience of generating / using docker images, so don't know. Would be good to discuss. I don't know if this is related, but I would like there to be some kind of pipeline for generating HTML that gets triggered whenever anyone pushes to remote so that latest doc can be easily accessed in a web browser.

@illsilin, would it make any sense to include doc-related deps to our docker images? If we run a github action to publish the docs, is it simpler to get all the deps in a docker image or install them as a part of a GHA?

**Comment by zjing14** [comment]
@geyyer Please review and approve this PR.

**Review by geyyer — APPROVED** [review]
LGTM!

**Pull request merged**